Fixup Extended Existential Metatype Casts

CodaFi opened the pull request:
The fix here is two-fold:

1) Teach SILGen that it cannot use the scalar casting paths for extended existentials
2) Teach the runtime casting entrypoint to unwrap as much parallel metatype structure as possible
   before arriving at a 'Self' type bound for the requirement checking paths.

The code here mirrors the destructuring check we're doing in remote mirrors.

rdar://95166916

CodaFi committed:
Fixup Extended Existential Metatype Casts

The fix here is two-fold:

1) Teach SILGen that it cannot use the scalar casting paths for extended existentials
2) Teach the runtime casting entrypoint to unwrap as much metatype structure as possible
   before arriving at a 'Self' type bound for the requirement checking paths.

The code here mirrors the destructuring check we're doing in remote mirrors.

rdar://95166916

CodaFi commented:
@swift-ci smoke test

CodaFi committed:
Failing Casts Should Not Crash

Return a NULL demangle tree instead of crashing. When an unconditional
runtime cast fails, it's going to crash anyways, but this way it prints
a better description (though not great) than the unreachable's message here.

CodaFi commented:
@swift-ci smoke test

CodaFi commented:
@swift-ci smoke test macOS

rjmccall commented on stdlib/public/runtime/DynamicCast.cpp:
```diff
@@ -1781,6 +1781,29 @@ static DynamicCastResult tryCastToExtendedExistential(
+
+      selfType = metatypeMetadata->InstanceType;
+      node = node->getChild(0)->getChild(0);
+    }
```
You should test that you got to a type parameter here or else bail out.

CodaFi committed:
Check The Last Node After Unwrapping Metatypes

Ensure the last node after all of the destructuring really
is a dependent generic param type. If it isn't, something has
gone horribly wrong.

rjmccall reviewed (approved):
Thanks, LGTM

CodaFi commented:
@swift-ci smoke test
